sdk/communication/AzureCommunication/Source/Authentication/HMACAuthenticationPolicy.swift

```diff
@@ -0,0 +1,144 @@
+    
+    public init(accessKey: String) {
+        self.accessKey = accessKey
+    }
+
+    public func authenticate(
+        request: PipelineRequest,
+        completionHandler: @escaping OnRequestCompletionHandler) {
+        let contents = request.httpRequest.data ?? Data() // Is this the body of the request?
+        
+        guard request.httpRequest.url.scheme?.contains("https") == true else {
+            completionHandler(
+                request,
+                AzureError.sdk("HMACAuthenticationPolicy requires a URL using the HTTPS protocol scheme"))
+            return
+        }
+        
+        request.httpRequest.headers = addAuthenticationHeaders(
+            url: request.httpRequest.url,
+            httpMethod: request.httpRequest.httpMethod.rawValue,
+            contents: contents)
+        
+        completionHandler(request, nil)
+    }
+    
+    public func addAuthenticationHeaders(
+        url: URL,
+        httpMethod: String,
+        contents: Data) -> HTTPHeaders {
+        var headers: HTTPHeaders = [:]
+        headers[HMACAuthenticationPolicy.contentHashHeader] = contents.sha256
+        
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "E, dd MMM YYYY HH:mm:ss 'GMT'" // Is this the right date format? Try ISO8601DateFormatter
+        dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
+        
+        let date = Date()
+        let utcNow = dateFormatter.string(from: date)
+        headers[HMACAuthenticationPolicy.dateHeader] = utcNow
+        headers[HMACAuthenticationPolicy.hostHeader] = url.host
+                
+        headers.merge(addSignatureHeader(
+                        url: url,
+                        httpMethod: httpMethod,
+                        date: utcNow,
+                        contentHashed: contents.sha256)) { (_, new) in new }
+        /**
+         After signature
+         HashMap@56 size=4
+         0:HashMap$Node@113 "date":"Wed, 07 Oct 2020 18:16:02 GMT"
+         1:HashMap$Node@114 "Authorization":"HMAC-SHA256 SignedHeaders=date;host;x-ms-content-sha256&Signature=KZD9UN4LsktsEX2e9cRp+LS2opjAtEVKqt+OzFCHh9o="
+         2:HashMap$Node@115 "x-ms-content-sha256":"YjVxGFu++f6tLM9YEVQVRmchZiYyxQ+8Bi3PXTJz2C4="
+         3:HashMap$Node@116 "host":"localhost"
+         */
+        return headers
+    }
+    
+    private func addSignatureHeader(
+        url: URL,
+        httpMethod: String,
+        date: String,
+        contentHashed: String) -> HTTPHeaders {
+        // Order of the headers are important here for generating correct signature
+        let signedHeaderNames = "\(HMACAuthenticationPolicy.dateHeader);\(HMACAuthenticationPolicy.hostHeader);\(HMACAuthenticationPolicy.contentHashHeader)"
+        let signedHeaderValues =  "\(date);\(url.host ?? "");\(contentHashed)" // date;urlhost;content hash
+        
+        // Add unit test for different ports
+        // 1 with port 1 without port
```

Comment:
does `port` include in the url.path?

Reply:
No. Port is returned under `url.port`. This was a suggestion from Dominik to write unit tests in the case different ports are used so I will have those in an update